langlib/lang.__internal/src/main/java/org/ballerinalang/langlib/internal/GetElementType.java

```diff
@@ -36,22 +36,27 @@
  */
 public class GetElementType {
 
+    private GetElementType() {
+    }
+
     public static BTypedesc getElementType(Object td) {
         BTypedesc bTypedesc = (BTypedesc) td;
-        return getTypeDescValue(bTypedesc.getDescribingType());
+        return getElementTypeDescValue(bTypedesc.getDescribingType());
     }
 
-    private static BTypedesc getTypeDescValue(Type type) {
-        if (type.getTag() == TypeTags.ARRAY_TAG) {
-            return ValueCreator.createTypedescValue(((ArrayType) type).getElementType());
-        } else if (type.getTag() == TypeTags.STREAM_TAG) {
-            return ValueCreator.createTypedescValue(((StreamType) type).getConstrainedType());
-        } else if (type.getTag() == TypeTags.TABLE_TAG) {
-            return ValueCreator.createTypedescValue(((TableType) type).getConstrainedType());
-        } else if (type.getTag() == TypeTags.FINITE_TYPE_TAG && ((FiniteType) type).getValueSpace().size() == 1) {
-            return getTypeDescValue(((BValue) (((FiniteType) type).getValueSpace().iterator().next())).getType());
+    private static BTypedesc getElementTypeDescValue(Type type) {
+        switch (type.getTag()) {
+            case TypeTags.ARRAY_TAG:
+                return ValueCreator.createTypedescValue(((ArrayType) type).getElementType());
+            case TypeTags.TUPLE_TAG:
+                return ValueCreator.createTypedescValue(
+                        TypeCreator.createUnionType(((TupleType) type).getTupleTypes()));
+            case TypeTags.FINITE_TYPE_TAG:
+                // this is reached only for immutable values
+                return getElementTypeDescValue(
+                        ((BValue) (((FiniteType) type).getValueSpace().iterator().next())).getType());
```

Comment:
It seems that due to the recent changes to the beta 3 branch read-only values also goes to other 3 cases.